src/generator/extract_schema.cpp

```diff
@@ -0,0 +1,323 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+#include <algorithm>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+#include <string_view>
+#include <unordered_set>
+#include <variant>
+
+#include "generator/extract_schema.hpp"
+
+namespace ddwaf::generator {
+namespace schema {
+
+struct node_record;
+struct node_array;
+
+using node_record_ptr = std::unique_ptr<node_record>;
+using node_array_ptr = std::unique_ptr<node_array>;
+
+enum class scalar_type : uint8_t { null = 1, boolean = 2, integer = 4, string = 8, real = 16 };
+
+struct node_scalar {
+    scalar_type type{scalar_type::null};
+    std::map<std::string_view, std::string_view> tags{};
+};
+
+using base_node = std::variant<std::monostate, node_scalar, node_array_ptr, node_record_ptr>;
+
+struct node_hash {
+    constexpr std::size_t operator()(const std::monostate & /*node*/) const noexcept
+    {
+        return 0x9e3779b9;
+    }
+    std::size_t operator()(const node_scalar &node) const noexcept;
+    std::size_t operator()(const node_array_ptr &node) const noexcept;
+    std::size_t operator()(const node_record_ptr &node) const noexcept;
+    std::size_t operator()(const base_node &node) const noexcept;
+};
+
+struct node_equal {
+    constexpr bool operator()(const std::monostate & /*lhs*/, const std::monostate & /*rhs*/) const
+    {
+        return true;
+    }
+    bool operator()(const node_scalar &lhs, const node_scalar &rhs) const;
+    bool operator()(const node_array_ptr &lhs, const node_array_ptr &rhs) const;
+    bool operator()(const node_record_ptr &lhs, const node_record_ptr &rhs) const;
+    template <typename T, typename U>
+    constexpr bool operator()(const T & /*lhs*/, const U & /*rhs*/) const
+        requires(!std::is_same_v<T, U>)
+    {
+        return false;
+    }
+    bool operator()(const base_node &lhs, const base_node &rhs) const;
+};
+
+struct node_record {
+    bool truncated{false};
+    std::map<std::string_view, base_node> children{};
```

Comment:
Mainly because comparing the equality of two maps is much simpler than comparing two unordered_maps, note that this is only done when inserting into a `node_array` (as these are actually supposed to be sets).

Reply:
What do you mean with simpler? More efficient? Maybe, as you can compare the elements in the same position in the tree instead of doing a lookup in the second hash table.

Incidentally, that leads me to think there is a bug here. std::map uses the equality function of `base_node`, not your `node_equal`.

Reply:
`unordered_map` was a tiny bit faster, so I've changed it.